src/serialize.rs

```diff
@@ -1,57 +1,80 @@
 
 use crate::allocator::{Allocator, NodePtr, SExp};
 use crate::node::Node;
+use crate::reduction::EvalErr;
 
 const MAX_SINGLE_BYTE: u8 = 0x7f;
 const CONS_BOX_MARKER: u8 = 0xff;
 
-fn bad_encoding() -> std::io::Error {
-    Error::new(ErrorKind::InvalidInput, "bad encoding")
+fn bad_encoding() -> io::Error {
+    io::Error::new(ErrorKind::InvalidInput, "bad encoding")
 }
 
-fn internal_error() -> std::io::Error {
-    Error::new(ErrorKind::InvalidInput, "internal error")
+fn internal_error() -> io::Error {
+    io::Error::new(ErrorKind::InvalidInput, "internal error")
 }
 
-fn encode_size(f: &mut dyn Write, size: u64) -> std::io::Result<()> {
-    if size < 0x40 {
-        f.write_all(&[(0x80 | size) as u8])?;
+/// write the correct prefix for an atom of size `size` whose first byte is `atom_0`.
+/// If the atom is of size 0, use any placeholder first byte, as it's ignored anyway.
+
+fn write_atom_encoding_prefix_with_size(
+    f: &mut dyn io::Write,
+    atom_0: u8,
+    size: u64,
+) -> io::Result<()> {
+    if size == 0 {
+        f.write_all(&[0x80])
+    } else if size == 1 && atom_0 < 0x80 {
+        Ok(())
+    } else if size < 0x40 {
+        f.write_all(&[0x80 | (size as u8)])
     } else if size < 0x2000 {
-        f.write_all(&[(0xc0 | (size >> 8)) as u8, ((size) & 0xff) as u8])?;
+        f.write_all(&[0xc0 | (size >> 8) as u8, size as u8])
     } else if size < 0x10_0000 {
         f.write_all(&[
             (0xe0 | (size >> 16)) as u8,
             ((size >> 8) & 0xff) as u8,
             ((size) & 0xff) as u8,
-        ])?;
+        ])
     } else if size < 0x800_0000 {
         f.write_all(&[
             (0xf0 | (size >> 24)) as u8,
             ((size >> 16) & 0xff) as u8,
             ((size >> 8) & 0xff) as u8,
             ((size) & 0xff) as u8,
-        ])?;
+        ])
     } else if size < 0x4_0000_0000 {
         f.write_all(&[
             (0xf8 | (size >> 32)) as u8,
             ((size >> 24) & 0xff) as u8,
             ((size >> 16) & 0xff) as u8,
             ((size >> 8) & 0xff) as u8,
             ((size) & 0xff) as u8,
-        ])?;
+        ])
     } else {
-        return Err(Error::new(ErrorKind::InvalidData, "atom too big"));
+        Err(io::Error::new(ErrorKind::InvalidData, "atom too big"))
     }
-    Ok(())
 }
 
-pub fn node_to_stream(node: &Node, f: &mut dyn Write) -> std::io::Result<()> {
+/// all atoms serialize their contents verbatim. All expect those one-byte atoms
+/// from 0x00-0x7f also have a prefix encoding their length. This function
+/// encodes that length prefix.
+
+fn write_atom_encoding_prefix(f: &mut dyn io::Write, atom: &[u8]) -> io::Result<()> {
+    let u8_0 = if !atom.is_empty() { atom[0] } else { 0 };
+    write_atom_encoding_prefix_with_size(f, u8_0, atom.len() as u64)
+}
+
+/// serialize an atom
+fn write_atom(f: &mut dyn io::Write, atom: &[u8]) -> io::Result<()> {
+    write_atom_encoding_prefix(f, atom)?;
+    f.write_all(atom)
+}
```

Comment:
these two functions should have unit tests, with the interesting edge cases.

Reply:
I've added tests (and removed a no-longer necessary intermediate function).